### PR TITLE
Refactoring and performance improvement on Cube2Equirec

### PR DESCRIPTION
**Pull request by matsuren**
## Motivation and Context
Thank you, @Skylion007 for implementing Cube2Equirec!
I refactored the source code to improve the performance. (I think it will also improve the readablity.)
Unfortunately, it will cause a confliction with #484. Please tell me how I should deal with the confliction.

## How Has This Been Tested
The source code was tested by the following code.
```python
#!/usr/bin/env python
# coding: utf-8
import habitat
from habitat_baselines.common.baseline_registry import baseline_registry
from habitat_baselines.utils.common import batch_obs
from habitat_baselines.common.obs_transformers import Cube2Equirec
import math
from copy import deepcopy
import matplotlib.pyplot as plt
from time import time
import Cube2Equirec_pullreq # Pull request model

equ_h = 256
equ_w = 512

config = habitat.get_config("configs/tasks/pointnav.yaml")
config.defrost()
CAMERA_NUM = 6
orient = [
    [0, math.pi, 0],  # Back
    [-math.pi / 2, 0, 0],  # Down
    [0, 0, 0],  # Front
    [0, math.pi / 2, 0],  # Right
    [0, 3 / 2 * math.pi, 0],  # Left
    [math.pi / 2, 0, 0],  # Up
]
sensor_uuids = []
# Setup six cameras, Back, Down, Front, Left, Right, Up.
if "RGB_SENSOR" in config.SIMULATOR.AGENT_0.SENSORS:
    config.SIMULATOR.RGB_SENSOR.ORIENTATION = orient[0]
    for camera_id in range(CAMERA_NUM):
        camera_template = f"RGB_{camera_id}"
        camera_config = deepcopy(config.SIMULATOR.RGB_SENSOR)
        camera_config.ORIENTATION = orient[camera_id]

        camera_config.UUID = camera_template.lower()
        sensor_uuids.append(camera_config.UUID)
        setattr(config.SIMULATOR, camera_template, camera_config)
        config.SIMULATOR.AGENT_0.SENSORS.append(camera_template)
config.freeze()

# Setup CubeMap2Equirec
obs_trans = baseline_registry.get_obs_transformer("CubeMap2Equirec")
cube2equirec = obs_trans(sensor_uuids, (equ_h, equ_w), camera_config["HEIGHT"])

# Load embodied AI task (PointNav) and a pre-specified virtual robot
env = habitat.Env(config=config)

observations = env.reset()
batch = batch_obs([observations])

# Prepare cubemap
sensor_obs = [batch[sensor] for sensor in sensor_uuids]
sensor_dtype = sensor_obs[0].dtype
# Stacking along axis makes the flattening go in the right order.
imgs = torch.stack(sensor_obs, axis=1)
imgs = torch.flatten(imgs, end_dim=1)
if not cube2equirec.channels_last:
    imgs = imgs.permute((0, 3, 1, 2))  # NHWC => NCHW
imgs = imgs.float()

# Create master and pull request model
master_model = Cube2Equirec(equ_h, equ_w, camera_config["HEIGHT"])
pullreq_model = Cube2Equirec_pullreq(equ_h, equ_w, camera_config["HEIGHT"])

device = torch.device("cpu")
imgs = imgs.to(device)
master_model = master_model.to(device)
pullreq_model = pullreq_model.to(device)

# master CPU performance
elps = []
for _ in range(1000):
    start = time()
    out = master_model(imgs)
    elps.append(time() - start)
print(f"Master elps [CPU]:{1000*sum(elps)/len(elps):.4f} [ms]")

# pullreq CPU performance
elps = []
for _ in range(1000):
    start = time()
    out = pullreq_model(imgs)
    elps.append(time() - start)
print(f"Pullreq elps [CPU]:{1000*sum(elps)/len(elps):.4f} [ms]")

device = torch.device("cuda")
imgs = imgs.to(device)
master_model = master_model.to(device)
pullreq_model = pullreq_model.to(device)

# master GPU performance
elps = []
out_master = master_model(imgs)
for _ in range(1000):
    start = time()
    out = master_model(imgs)
    torch.cuda.synchronize()
    elps.append(time() - start)
print(f"Master elps [GPU]:{1000*sum(elps)/len(elps):.4f} [ms]")

# pullreq GPU performance
elps = []
out_pullreq = pullreq_model(imgs)
for _ in range(1000):
    start = time()
    out = pullreq_model(imgs)
    torch.cuda.synchronize()
    elps.append(time() - start)
print(f"Pullreq elps [GPU]:{1000*sum(elps)/len(elps):.4f} [ms]")


# Visualization
out_master = out_master.to(dtype=sensor_dtype)
out_master = out_master.permute((0, 2, 3, 1))  # NCHW => NHWC
out_master = out_master.cpu().numpy().squeeze()

out_pullreq = out_pullreq.to(dtype=sensor_dtype)
out_pullreq = out_pullreq.permute((0, 2, 3, 1))  # NCHW => NHWC
out_pullreq = out_pullreq.cpu().numpy().squeeze()

fig, ax = plt.subplots(2, 1, figsize=(8, 8))
ax[0].imshow(out_master)
ax[0].set_title("Master")
ax[1].imshow(out_pullreq)
ax[1].set_title("PullReq")
fig.set_facecolor("white")
plt.show()
```
And, the performance of each model:
```
Master elps [CPU]:23.0302 [ms]
Pullreq elps [CPU]:4.1277 [ms]
Master elps [GPU]:0.9772 [ms]
Pullreq elps [GPU]:0.1641 [ms]
```
The generated equirectangular images:
Figure_1

Tested environments:
 - CPU:Intel® Core™ i7-6800K CPU
 - GPU:TITAN X (Pascal) 

## Types of changes

- refactoring
- performance improvement

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

Thank you in advance for reviewing this pull request!

### CODE REVIEW

File: habitat_baselines/common/obs_transformers.py [hunk 1]
```diff
@@ -482,6 +303,23 @@ def to_equirec_tensor(self, batch: torch.Tensor):
...
 
+    # Get theta and phi map
+    def get_theta_phi_map(self, equ_h: int, equ_w: int):
+        phi, theta = torch.meshgrid(torch.arange(equ_h), torch.arange(equ_w))
+        theta_map = -(theta + 0.5) * 2 * np.pi / equ_w + np.pi
+        phi_map = -(phi + 0.5) * np.pi / equ_h + np.pi / 2
+        return theta_map, phi_map
+
+    # Project on unit sphere
+    def angle2sphere(self, theta_map: torch.Tensor, phi_map: torch.Tensor):
```
Comment (Skylion007) [inline habitat_baselines/common/obs_transformers.py:314]: Missing return type:

File: habitat_baselines/common/obs_transformers.py [hunk 2]
```diff
@@ -482,6 +303,23 @@ def to_equirec_tensor(self, batch: torch.Tensor):
     def forward(self, batch: torch.Tensor):
         return self.to_equirec_tensor(batch)
 
+    # Get theta and phi map
+    def get_theta_phi_map(self, equ_h: int, equ_w: int):
```
Comment (Skylion007) [inline habitat_baselines/common/obs_transformers.py:307]: Missing return type:

File: habitat_baselines/common/obs_transformers.py [hunk 3]
```diff
@@ -224,247 +223,69 @@ class Cube2Equirec(nn.Module):
     optimized for modern PyTorch."""
 
     def __init__(
-        self, equ_h: int, equ_w: int, cube_length: int, fov: int = 90
+        self, equ_h: int, equ_w: int, cube_length=None, fov: int = 90
```
Comment (Skylion007) [inline habitat_baselines/common/obs_transformers.py:226]: missing typehint

**Comment by Skylion007** [comment]
@matsuren I just implemented a similar performance improvement in #484, but I further improved it by removing the loop for batched input as well. Can you compare performance to that PR / integrate the .view() tricks I did there to further improve performance?

**Review by Skylion007 — CHANGES_REQUESTED** [review]
Integrate with the performance improvements made in #484

File: habitat_baselines/common/obs_transformers.py [hunk 4]
```diff
@@ -224,247 +223,69 @@ class Cube2Equirec(nn.Module):
...
 
     # Convert input cubic tensor to output equirectangular image
     def to_equirec_tensor(self, batch: torch.Tensor):
         # Move the params to the right device. NOOP after first call
-        self.grid = self.grid.to(batch.device)
-        self.orientation_mask = self.orientation_mask.to(batch.device)
+        self.grids = self.grids.to(batch.device)
         # Check whether batch size is 6x
         batch_size = batch.size()[0]
         if batch_size % 6 != 0:
```
Comment (Skylion007) [inline habitat_baselines/common/obs_transformers.py:291]: Remove the for loop here and replace it with the changes made in #484

File: habitat_baselines/common/obs_transformers.py [hunk 5]
```diff
@@ -224,247 +223,69 @@ class Cube2Equirec(nn.Module):
...
+            grid[
+                ~mask
+            ] = 2  # values bigger than one will be ignored by grid_sample
+            grid_xy = -grid[..., :2].unsqueeze(0)
+            grids.append(grid_xy)
+        grids = torch.cat(grids, dim=0)
+        return grids
 
     # Convert cubic images to equirectangular
     def _to_equirec(self, batch: torch.Tensor):
```
Comment (Skylion007) [inline habitat_baselines/common/obs_transformers.py:275]: Whoops, i missing a return type here, should fix that.

File: habitat_baselines/common/obs_transformers.py [hunk 6]
```diff
@@ -224,264 +223,108 @@ class Cube2Equirec(nn.Module):
...
-
-        output = torch.zeros(
-            1, ch, self.equ_h, self.equ_w, device=batch.device
+        cubemap_sides = batch.size()[0]
+        if cubemap_sides == 0 or cubemap_sides % 6 != 0:
+            raise ValueError("Batch size should be 6x")
+        output = torch.nn.functional.grid_sample(
+            batch,
+            self.grids.repeat(batch_size // 6, 1, 1, 1),
+            align_corners=True,
```
Comment (Skylion007) [inline habitat_baselines/common/obs_transformers.py:287]: Align_corners here is nice, but it does import a performance penalty if I recall, is it worth it?

File: habitat_baselines/common/obs_transformers.py [hunk 7]
```diff
@@ -223,265 +222,100 @@ class Cube2Equirec(nn.Module):
...
```
Comment (Skylion007) [inline habitat_baselines/common/obs_transformers.py:278]: Are you sure we should use align_corners=True? https://github.com/pytorch/pytorch/issues/20785

Comment (Skylion007) [reply habitat_baselines/common/obs_transformers.py:278]: Ah, nvm. We need that for the optimized CUDA kernels.

**Review by Skylion007 — APPROVED** [review]

**Comment by matsuren** [comment]
Thank you so much for taking care of my pull request! It looks good to me. Sorry about some missing type hinting. It's my first time to use them. Next time I will be careful about that. 

I'm thinking about implementing for fisheye. Once I finish it, I might send pull request. 

Thanks!

**Pull request merged**